### PR TITLE
multi: tweaks to fix gcs indexes

### PR DESCRIPTION
**Pull request by tuxcanfly**
This PR contains misc changes to help in using this library in bcoin to support BIP 158.

Ref: https://github.com/bcoin-org/bcoin/pull/370

### CODE REVIEW

**Comment by tuxcanfly** [comment]
* Updated Filter Hash, Header as per reference:

  - https://github.com/Roasbeef/btcutil/blob/master/gcs/builder/builder.go#L373-L382

```
// GetFilterHash returns the double-SHA256 of the filter.
func GetFilterHash(filter *gcs.Filter) chainhash.Hash {
	var zero chainhash.Hash
	if filter == nil {
		return zero
	}

	hash1 := chainhash.HashH(filter.NBytes())
	return chainhash.HashH(hash1[:])
}
```

* According to BIP 158:
 
  - `Each data push in the scriptPubKey of each output, only if the scriptPubKey is parseable`

  - Updated to handle error compiling the script and return.

*  `OP_0` empty buffer is considered as pushed data and indexed. Updated to handle the same.

* When creating extended filters, it's easier to check `data.length` than catch an exception

  - Updated assert to allow 0 length items

**Comment by tuxcanfly** [comment]
Also updated not to ignore any empty push datas. Will update with tests.

File: lib/golomb.js [hunk 1]
```diff
@@ -346,8 +349,15 @@ function compare(a, b) {
 }
 
 function getPushes(items, script) {
+  try {
+    script.compile();
```
Comment (chjj) [inline lib/golomb.js:353]: Why is this necessary? This will hurt performance greatly.

**Comment by chjj** [comment]
Will merge once `script.compile()` is removed.

File: lib/golomb.js (hunk 1 above)
Comment (tuxcanfly) [reply lib/golomb.js:353]: @chjj IIRC data pushes from some invalid scripts were still being added to the items to be indexed, causing a fork in the filter chain. I can dig into the exact block but OTOH it was around 10k height in testnet.

Comment (tuxcanfly) [reply lib/golomb.js:353]: as per the BIP: Each data push in the scriptPubKey of each output, *only if* the scriptPubKey is parseable

Comment (chjj) [reply lib/golomb.js:353]: @tuxcanfly, we can check for that a quicker way. The scripting system tracks bad pushdatas with `op.value === -1`. I'll merge this and add the check.

**Comment by tuxcanfly** [comment]
@chjj thanks. You can check with the testnet block

`000000000e09b2c50f124858173313fbe8d47f69a660a5fe209598dc07b02ade`

expected filter starts with `8a 4f 78 49 e3 2a 71 3e`

**Comment by chjj** [comment]
Publishing as v0.2.0. Might need to update a package.json.

**Comment by tuxcanfly** [comment]
Great, that works. Will update the dependencies for packages, thanks.
